extensions/oidc-client/runtime/src/main/java/io/quarkus/oidc/client/runtime/GrantTokensProducer.java

```diff
@@ -0,0 +1,15 @@
+package io.quarkus.oidc.client.runtime;
+
+import javax.enterprise.context.RequestScoped;
+import javax.enterprise.inject.Produces;
+
+import io.quarkus.oidc.client.GrantTokens;
+
+@RequestScoped
+public class GrantTokensProducer extends AbstractGrantTokensProducer {
+
+    @Produces
+    public GrantTokens produceTokens() {
```

Comment:
@stuartwdouglas I'll check, sure, I have a test checking the logs which confirm the tokens have been refreshed (with the filter which uses `@OidcClient`) and I have a test which verifies the tokens have been refreshed with `@GrantTokensProducer/@GrantTokens` but I'll add the log check there as well :-)

Reply:
Your test works because you have explicitly marker the filter as @RequestScoped. But this means that every request will require a request to KC to get the keys. 

I think the keys should be stored in a singleton, which will mean that most requests don't need to make any requests to KC. When they keys expire the singleton can manage the refresh process and make sure they are only refreshed once.

Reply:
@stuartwdouglas Oh, do you mean the filter is re-initialized on every request ? that is not good then yeah